rj_gameplay/stp/role/assignment/naive.py

```diff
@@ -0,0 +1,164 @@
+from collections import defaultdict
+from math import isfinite
+from typing import Tuple, List
+
+import scipy.optimize
+import numpy as np
+
+import stp.role as role
+import stp.role.assignment as assignment
+import stp.game_state
+
+SortedRequests = List[assignment.FlatRoleRequests]
+
+
+class NaiveRoleAssignment(assignment.IRoleAssignment):
+    @staticmethod
+    def get_sorted_requests(requests: assignment.FlatRoleRequests) -> SortedRequests:
+        """Returns a list of FlatRoleRequests sorted in ascending priority order.
+        :param requests:
+        :return:
+        """
+        role_id: assignment.RoleId
+        request: role.RoleRequest
+
+        sorted_requests: SortedRequests = [
+            {} for _ in range(role.Priority.NUM_PRIORITIES)
+        ]
+
+        for role_id, request in requests.items():
+            sorted_requests[int(request.priority)][role_id] = request
+
+        return sorted_requests
+
+    @staticmethod
+    def compute_costs_matrix(
+        free_robots: np.ndarray,
+        flat_requests: assignment.FlatRoleRequests,
+        game_state: stp.game_state.GameState,
+        prev_results: assignment.FlatRoleResults,
+    ) -> np.ndarray:
+        """Computes the m x n cost matrix corresponding to the passed in free robots and
+        role requests.
+        :param flat_requests: Role requests to compute cost matrix for.
+        :param free_robots: Free robots to compute cost matrix for.
+        :param game_state: Current game state.
+        :param prev_results: The previous results.
+        :return: The m x n cost matrix corresponding to the passed in free robots and
+        role requests.
+        """
+
+        # Cost matrix size (n_robots, n_requests)
+        robot_costs: np.ndarray = np.zeros((len(free_robots), len(flat_requests)))
+
+        # Iterate over each role request.
+        for request_idx, (role_id, request) in enumerate(flat_requests.items()):
+            role_id: assignment.RoleId
+            request: role.RoleRequest
+
+            # For each role request, iterate over robots.
+            for robot_idx, robot in enumerate(free_robots):
+                # Otherwise, record the cost.
+                cost: float = request.cost_fn(robot, prev_results.get(role_id, None))
+                if not isfinite(cost):
```

Comment:
I think we were planning to express hard constraints as a separate `bool` return value. Infinities are annoying because inevitably someone will try to combine two costs and forget to deal with that case, making it explicit makes it impossible to express.

If there is a `NaN` in the cost function, role assignment should be noisy and throw an exception.

That said, is a `0` the right thing to do in this case? Because that's what you're currently putting into the cost matrix, which seems _really_ wrong.

Reply:
The `isfinite` here was mainly to handle bad cost functions, ie trying to divide by 0 or something, rather than hard constraints. This still needs to be worked on, but I just thought that it would be nice if role assignment still worked, but it gets blacklisted + a log message or something.

Reply:
Ok. Yeah honestly I think we should just throw for now.